x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/time/RemainingTime.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.common.time;
+
+import org.elasticsearch.core.TimeValue;
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.function.Supplier;
+
+public interface RemainingTime extends Supplier<TimeValue> {
+    /**
+     * Create a {@link Supplier} that returns a decreasing {@link TimeValue} on each invocation, representing the amount of time until
+     * the call times out.  The timer starts when this method is called and counts down from remainingTime to 0.
+     * currentTime should return the most up-to-date system time, for example Instant.now() or Clock.instant().
+     */
+    static RemainingTime from(Supplier<Instant> currentTime, TimeValue remainingTime) {
+        var timeout = currentTime.get().plus(remainingTime.duration(), remainingTime.timeUnit().toChronoUnit());
+        var maxRemainingTime = remainingTime.nanos();
+        return () -> {
+            var remainingNanos = ChronoUnit.NANOS.between(currentTime.get(), timeout);
+            return TimeValue.timeValueNanos(Math.max(0, Math.min(remainingNanos, maxRemainingTime)));
```

Comment:
Is the logic with `maxRemainingTime` necessary when `maxRemainingTime == remainingNanos`?
Is it to handle the case where `currentTime` goes backwards?

Reply:
I saw one case where `currentTime` went backwards.  There's a monotonic clock in Transforms that we could move to commons to help prevent this, but for now I just set a max time in hopes that it doesn't happen that often :crossed_fingers: